engine/pkg/externalresource/model/model.go

```diff
@@ -165,10 +168,31 @@ func ParseResourceID(rpath ResourceID) (ResourceType, ResourceName, error) {
 	}
 
 	suffix := path.Join(segments[1:]...)
-	return resourceType, suffix, nil
+	return resourceType, EncodeResourceName(suffix), nil
 }
 
 // BuildResourceID returns an ResourceID based on given ResourceType and ResourceName.
-func BuildResourceID(rtype ResourceType, name ResourceName) ResourceID {
+func BuildResourceID(rtype ResourceType, resName ResourceName) ResourceID {
+	name, err := DecodeResourceName(resName)
+	if err != nil {
+		log.Panic("invalid resource name", zap.Error(err))
+	}
 	return path.Join("/"+string(rtype), name)
 }
+
+// EncodeResourceName encodes raw resource name to a valid resource name.
+func EncodeResourceName(resNameStr string) ResourceName {
```

Comment:
why we need to repeat `resNameStr` twice in the return value?

Reply:
This is used for debugging and only the part of hex encode is kept in the design